File: src/types/geoRisk.ts
/** TypeScript types for Geopolitical Health Scan - mirrors backend schema */

export type RiskTolerance = "low" | "medium" | "high";
export type Confidence = "low" | "medium" | "high";
export type ScenarioName = "low" | "moderate" | "severe";

export type AssetClass = 
  | "Equities"
  | "Fixed Income"
  | "Alternatives"
  | "Cash"
  | "Commodities"
  | "Real Estate"
  | "Private Equity";

export interface Holding {
  id: string;
  name: string;
  ticker?: string;
  isin?: string;
  country?: string;
  class: AssetClass;
  sector: string;
  region: string;
  sub_region?: string;
  quantity?: number; // Shares/units held
  value: number; // Current market value
  cost_basis?: number; // Original investment amount
  allocation_pct: number; // % of portfolio
  allocation_class_pct?: number; // % of asset class
  pnl_to_date: number; // Profit/Loss to date
  pnl_pct: number; // Return % since entry
  ytd_return?: number; // Year-to-date return %
  last_price?: number; // Current market price per unit
  last_valuation_date?: string; // YYYY-MM-DD
  entry_date?: string; // YYYY-MM-DD
  currency: string;
  beta?: number; // Market correlation (for equities)
  duration?: number; // Interest rate sensitivity (for bonds)
  credit_rating?: string; // Credit quality (for bonds)
  gp_score?: {
    sell: number; // Probability (0-1)
    hold: number; // Probability (0-1)
    buy: number; // Probability (0-1)
  };
}

export interface Portfolio {
  total_value: number;
  holdings: Holding[];
}

export interface GeoRiskScanInputs {
  client_id: string;
  as_of: string; // YYYY-MM-DD
  horizon_days: number;
  risk_tolerance: RiskTolerance;
  portfolio: Portfolio;
}

export interface Scenario {
  name: ScenarioName;
  p: number; // 0-1
}

export interface Citation {
  doc_id: string;
  snippet_id: string;
  section: string;
}

export interface GeoRiskOutput {
  scenarios: Scenario[];
  confidence: Confidence;
  drivers: string[];
  suitability_impact: string;
  limitations: string[];
  disclaimer: string;
  citations: Citation[];
}

export interface ValidationResult {
  passed: boolean;
  errors: string[];
}

export interface GeoRiskScanMeta {
  model: string;
  used_fallback: boolean;
  validation: ValidationResult;
}

export interface GeoRiskScanResult {
  scan_id: string;
  created_at: string; // ISO8601
  inputs: GeoRiskScanInputs;
  geo_risk: GeoRiskOutput;
  meta: GeoRiskScanMeta;
}

export interface GeoRiskScanRequest {
  client_id: string;
  as_of: string; // YYYY-MM-DD
  horizon_days: number;
  risk_tolerance: RiskTolerance;
  portfolio: Portfolio;
}

// Detailed pipeline result types
export interface ThemeDetail {
  theme: string;
  relevance_score: number;
  reasoning: string;
  keywords_matched: string[];
}

export interface IntelligenceSignalDetail {
  source: string;
  title: string;
  summary: string;
  topic: string;
  relevance_score: number;
  theme_match: string | null;
  published_at: string;
  url?: string | null;
  country?: string | null;
  activity_level?: string | null;
  base_relevance: number;
  theme_match_score: number;
  recency_score: number;
  source_quality: number;
  activity_level_score: number;
}

export interface ThemeImpactDetail {
  theme: string;
  impact_direction: string;
  impact_magnitude: number;
  confidence: number;
  reasoning: string;
  signal_count: number;
  summary?: string;
}

export interface AggregateImpactDetail {
  overall_direction: string;
  overall_magnitude: number;
  confidence: number;
  theme_impacts: ThemeImpactDetail[];
  total_signals: number;
}

export interface ActionProbabilitiesDetail {
  sell: number;
  hold: number;
  buy: number;
}

export interface WebSearchDetail {
  theme: string;
  query: string;
  results_count: number;
  signals_count: number;
  error?: string | null;
}

export interface DetailedPipelineResult {
  characterization_summary: string;
  asset_country: string | null;
  asset_region: string;
  asset_sub_region: string | null;
  asset_type: string;
  asset_class: string;
  asset_sector: string;
  is_emerging_market: boolean;
  is_developed_market: boolean;
  is_global_fund: boolean;
  exposures: string[];
  themes: ThemeDetail[];
  top_themes: string[];
  signals: IntelligenceSignalDetail[];
  signal_count: number;
  web_searches: WebSearchDetail[];
  impact: AggregateImpactDetail;
  probabilities: ActionProbabilitiesDetail;
  probability_summary: string;
  risk_tolerance: string;
  days_lookback: number;
}

File: backend/schemas/geo_risk.py
"""Pydantic schemas for Geopolitical Health Scan."""
from datetime import datetime
from typing import List, Literal
from pydantic import BaseModel, Field


class Holding(BaseModel):
    """Portfolio holding with detailed properties."""
    id: str
    name: str
    class_: str = Field(..., alias="class", description="Asset class: Equities, Fixed Income, Alternatives, etc.")
    region: str
    sector: str
    value: float
    allocation_pct: float = Field(..., ge=0.0, le=100.0, description="Percentage of portfolio")
    pnl_to_date: float = Field(default=0.0, description="Profit/Loss to date")
    pnl_pct: float = Field(default=0.0, description="PnL as percentage")
    currency: str = Field(default="USD")
    entry_date: str | None = Field(default=None, description="YYYY-MM-DD")
    last_valuation_date: str | None = Field(default=None, description="YYYY-MM-DD")
    ticker: str | None = Field(default=None, description="Ticker symbol for publicly traded securities")
    isin: str | None = Field(default=None, description="International Securities Identification Number")
    country: str | None = Field(default=None, description="Specific country if applicable")
    sub_region: str | None = Field(default=None, description="Sub-region e.g., Western Europe")
    gp_score: dict | None = Field(
        default=None,
        description="Geopolitical risk score with Sell/Hold/Buy probabilities"
    )
    
    class Config:
        populate_by_name = True


class Portfolio(BaseModel):
    """Client portfolio structure."""
    total_value: float
    holdings: List[Holding]


class GeoRiskScanInputs(BaseModel):
    """Inputs for a geopolitical risk scan."""
    client_id: str
    as_of: str = Field(..., description="YYYY-MM-DD format")
    horizon_days: int = Field(..., ge=1, le=3650)
    risk_tolerance: Literal["low", "medium", "high"]
    portfolio: Portfolio


class Scenario(BaseModel):
    """Risk scenario with probability."""
    name: Literal["low", "moderate", "severe"]
    p: float = Field(..., ge=0.0, le=1.0, description="Probability (0-1)")


class Citation(BaseModel):
    """Regulatory citation reference."""
    doc_id: str = Field(default="FCA_EXCERPTS")
    snippet_id: str
    section: str


class GeoRiskOutput(BaseModel):
    """Geopolitical risk assessment output."""
    scenarios: List[Scenario] = Field(..., min_length=3, max_length=3)
    confidence: Literal["low", "medium", "high"]
    drivers: List[str] = Field(..., min_length=1, max_length=10)
    suitability_impact: str = Field(..., min_length=50, max_length=500)
    limitations: List[str] = Field(..., min_length=1, max_length=5)
    disclaimer: str = "Internal decision-support only. Not financial advice."
    citations: List[Citation] = Field(default_factory=list)


class ValidationResult(BaseModel):
    """Validation metadata."""
    passed: bool
    errors: List[str] = Field(default_factory=list)


class GeoRiskScanMeta(BaseModel):
    """Metadata about the scan."""
    model: str
    used_fallback: bool = False
    validation: ValidationResult


class GeoRiskScanResult(BaseModel):
    """Complete geopolitical risk scan result."""
    scan_id: str
    created_at: str  # ISO8601 format
    inputs: GeoRiskScanInputs
    geo_risk: GeoRiskOutput
    meta: GeoRiskScanMeta


class GeoRiskScanRequest(BaseModel):
    """Request payload for POST /geo-risk/scan."""
    client_id: str
    as_of: str = Field(..., description="YYYY-MM-DD format")
    horizon_days: int = Field(default=365, ge=1, le=3650)
    risk_tolerance: Literal["low", "medium", "high"]
    portfolio: Portfolio


# Detailed pipeline response schemas
class ThemeDetail(BaseModel):
    """Theme identification detail."""
    theme: str
    relevance_score: float
    reasoning: str
    keywords_matched: List[str]


class IntelligenceSignalDetail(BaseModel):
    """Intelligence signal detail."""
    source: str
    title: str
    summary: str
    topic: str
    relevance_score: float
    theme_match: str | None
    published_at: str
    url: str | None = None
    country: str | None = None
    activity_level: str | None = None
    base_relevance: float = 0.0
    theme_match_score: float = 0.0
    recency_score: float = 0.0
    source_quality: float = 0.0
    activity_level_score: float = 0.0


class ThemeImpactDetail(BaseModel):
    """Theme impact assessment detail."""
    theme: str
    impact_direction: str
    impact_magnitude: float
    confidence: float
    reasoning: str
    signal_count: int
    summary: str = ""


class AggregateImpactDetail(BaseModel):
    """Aggregate impact assessment detail."""
    overall_direction: str
    overall_magnitude: float
    confidence: float
    theme_impacts: List[ThemeImpactDetail]
    total_signals: int


class ActionProbabilitiesDetail(BaseModel):
    """Action probabilities detail."""
    sell: float
    hold: float
    buy: float


class WebSearchDetail(BaseModel):
    """Web search metadata for a theme."""
    theme: str
    query: str
    results_count: int
    signals_count: int
    error: str | None = None


class DetailedPipelineResult(BaseModel):
    """Detailed pipeline result with all intermediate steps."""
    # Step 1: Characterization
    characterization_summary: str
    asset_country: str | None
    asset_region: str
    asset_sub_region: str | None
    asset_type: str
    asset_class: str
    asset_sector: str
    is_emerging_market: bool
    is_developed_market: bool
    is_global_fund: bool
    exposures: List[str]
    
    # Step 2: Theme identification
    themes: List[ThemeDetail]
    top_themes: List[str]
    
    # Step 3: Intelligence retrieval
    signals: List[IntelligenceSignalDetail]
    signal_count: int
    web_searches: List[WebSearchDetail]  # Web search metadata
    
    # Step 4: Impact assessment
    impact: AggregateImpactDetail
    
    # Step 5: Probability calculation
    probabilities: ActionProbabilitiesDetail
    probability_summary: str
    
    # Metadata
    risk_tolerance: str
    days_lookback: int
